thrift/binary.go

```diff
@@ -1292,6 +1293,394 @@ func (p *BinaryProtocol) DecodeText(val string, desc *TypeDescriptor, disallowUn
 	}
 }
 
+// GoType2ThriftType a go primitive type to a thrift type
+// The rules is:
+//   - bool -> BOOL
+//   - byte/int8 -> BYTE
+//   - int16 -> I16
+//   - int32 -> I32
+//   - int64/int -> I64
+//   - int -> I64
+//   - float64/float32 -> DOUBLE
+//   - string/[]byte -> STRING
+//   - []interface{} -> LIST
+//   - map[FieldID]interface{} -> STRUCT
+//   - map[(int|string|interface{})]interface{} -> MAP
+func GoType2ThriftType(val interface{}) (Type, error) {
+	_, ok := val.(map[FieldID]interface{})
+	if ok {
+		return STRUCT, nil
+	}
+	_, ok = val.([]byte)
+	if ok {
+		return STRING, nil
+	}
+	switch reflect.TypeOf(val).Kind() {
+	case reflect.Bool:
+		return BOOL, nil
+	case reflect.Int8, reflect.Uint8:
+		return BYTE, nil
+	case reflect.Int16, reflect.Uint16:
+		return I16, nil
+	case reflect.Int32, reflect.Uint32:
+		return I32, nil
+	case reflect.Int64, reflect.Uint64, reflect.Int, reflect.Uint:
+		return I64, nil
+	case reflect.Float64:
+		return DOUBLE, nil
+	case reflect.String:
+		return STRING, nil
+	case reflect.Slice:
+		return LIST, nil
+	case reflect.Map:
+		return MAP, nil
+	case reflect.Struct:
+		return STRUCT, nil
+	case reflect.Ptr:
+		return GoType2ThriftType(reflect.ValueOf(val).Elem().Interface())
+	default:
+		return STOP, errUnsupportedType
+	}
+}
+
+// ReadAny reads a thrift value from buffer and convert it to go primitive type
+// It basicallly obeys rules in `GoType2ThriftType`.
+// Specially,
+//   - For INT(8/16/32/64) type, the return type is corresponding int8/int16/int32/int64 by default;
+//   - For MAP type, the output key type could be string, int or interface{}, depends on the input key's thrift type.
+//   - for STRUCT type, the return type is map[thrift.FieldID]interface{}.
+func (p *BinaryProtocol) ReadAny(typ Type, strAsBinary bool, byteAsInt8 bool) (interface{}, error) {
+	switch typ {
+	case BOOL:
+		return p.ReadBool()
+	case BYTE:
+		if byteAsInt8 {
+			n, e := p.ReadByte()
+			return int8(n), e
+		}
+		return p.ReadByte()
+	case I16:
+		return p.ReadI16()
+	case I32:
+		return p.ReadI32()
+	case I64:
+		return p.ReadI64()
+	case DOUBLE:
+		return p.ReadDouble()
+	case STRING:
+		if strAsBinary {
+			return p.ReadBinary(false)
+		}
+		return p.ReadString(false)
+	case LIST, SET:
+		elemType, size, e := p.ReadListBegin()
+		if e != nil {
+			return nil, e
+		}
+		ret := make([]interface{}, 0, size)
```

Comment:
Vulnerable to maliciously constructed binary with an unreasonable large size. Same below.

Reply:
this is hard to prevent..